### PR TITLE
Rebase rvh squash 0912

### PR DESCRIPTION
**Pull request by jueshiwenli**

### CODE REVIEW

**Review by lqr8 — APPROVED** [review]

File: src/arch/riscv/regs/misc.hh [hunk 1]
```diff
@@ -851,7 +851,8 @@ const RegVal MSTATUS_MASK = STATUS_SD_MASK | STATUS_SXL_MASK |
                             STATUS_SPP_MASK | STATUS_MPIE_MASK |
                             STATUS_SPIE_MASK | STATUS_UPIE_MASK |
                             STATUS_MIE_MASK | STATUS_SIE_MASK |
-                            STATUS_UIE_MASK;
```
Comment (shinezyy) [inline src/arch/riscv/regs/misc.hh:665]: Delete old code?

File: src/arch/riscv/isa/formats/standard.isa [hunk 2]
```diff
@@ -345,7 +345,8 @@ def template CSRExecute {{
         RegVal data, olddata;
         auto lowestAllowedMode = (PrivilegeMode)bits(csr, 9, 8);
         auto pm = (PrivilegeMode)xc->readMiscReg(MISCREG_PRV);
-        if (pm < lowestAllowedMode) {
+        if ((pm < lowestAllowedMode)&&
+            ((csr <CSR_HSTATUS) ||(csr >CSR_HGATP)) && ((csr <CSR_VSSTATUS) ||(csr > CSR_VSATP))) {
```
Comment (shinezyy) [inline src/arch/riscv/isa/formats/standard.isa]: format thiese two line of code?

`csr <CSR_HSTATUS --> csr < CSR_HSTATUS`
`) ||( --> ) || (`

File: src/cpu/base.cc [hunk 3]
```diff
@@ -1114,6 +1114,232 @@ BaseCPU::diffWithNEMU(ThreadID tid, InstSeqNum seq)
...
+        if (gem5_val != ref_val) {
+            warn("Inst [sn:%lli] pc: %#lx\n", seq, diffInfo.pc->instAddr());
+            warn("Diff at \033[31m%s\033[0m Ref value: \033[31m"
+                    "%#lx\033[0m, GEM5 value: \033[31m%#lx\033[0m\n",
+                    "vsscratch", ref_val, gem5_val);
+            diffInfo.errorCsrsValue[CsrRegIndex::vsscratch] = 1;
+            diffAllStates->gem5RegFile.vsscratch = gem5_val;
+            if (!diff_at)
+                diff_at = ValueDiff;
+        }
```
Comment (shinezyy) [inline src/cpu/base.cc:1243]: Can we rewrite this chunk of code with a function, a template function, or a macro?

It looks too redundant

File: src/cpu/base.hh [hunk 4]
```diff
@@ -743,7 +743,7 @@ class BaseCPU : public ClockedObject
         uint64_t amoOldGoldenValue;
         // Register address causing difftest error
         bool errorRegsValue[96];// 32 regs + 32fprs +32 vprs
-        bool errorCsrsValue[32];// CsrRegIndex
+        bool errorCsrsValue[36];// CsrRegIndex
```
Comment (shinezyy) [inline src/cpu/base.hh]: Can we use a enum or macro here?

File: src/arch/riscv/isa/decoder.isa [hunk 5]
```diff
@@ -4316,22 +4316,54 @@ decode QUADRANT default Unknown::unknown() {
...
-                                            machInst);
-                                NPC = NPC;
-                            } else {
-                                xc->setMiscReg(MISCREG_PRV, status.spp);
-                                status.sie = status.spie;
-                                status.spie = 1;
-                                status.spp = PRV_U;
-                                xc->setMiscReg(MISCREG_STATUS, status);
-                                NPC = xc->readMiscReg(MISCREG_SEPC);
+                            bool writeSign = false;
```
Comment (shinezyy) [inline src/arch/riscv/isa/decoder.isa]: local_variable should use snake_case

File: src/arch/riscv/faults.cc [hunk 6]
```diff
@@ -98,6 +100,18 @@ RiscvFault::invoke(ThreadContext *tc, const StaticInstPtr &inst)
...
+        if (isInterrupt()) {
+            if (pp != PRV_M && (tc->readMiscReg(MISCREG_VIRMODE) == 1) &&
+                bits(tc->readMiscReg(MISCREG_HIDELEG), _code) != 0) {
+                prvh = PRV_HS;
+            }
+        } else {
+            if (pp != PRV_M && (tc->readMiscReg(MISCREG_VIRMODE) == 1) &&
+                bits(tc->readMiscReg(MISCREG_HEDELEG), _code) != 0) {
+                prvh = PRV_HS;
+            }
```
Comment (shinezyy) [inline src/arch/riscv/faults.cc:117]: the if block and else block seems to be the same?

**Review by shinezyy — CHANGES_REQUESTED** [review]

File: src/arch/riscv/pagetable_walker.cc [hunk 7]
```diff
@@ -979,7 +1259,15 @@ Walker::WalkerState::recvPacket(PacketPtr pkt)
         nextState = Ready;
         PacketPtr write = NULL;
         read = pkt;
-        mainFault = stepWalk(write);
+        if ((translateMode == twoStageMode) && (inGstage)){
+            mainFault = twoStageStepWalk(write);
+        }
+        else if ((translateMode == twoStageMode) && (!inGstage)){
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc]: } else if {

File: src/arch/riscv/pagetable_walker.cc [hunk 8]
```diff
@@ -979,7 +1259,15 @@ Walker::WalkerState::recvPacket(PacketPtr pkt)
...
+            mainFault = twoStageWalk(write);
+        }
+        else{
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc]: } else if {

File: src/arch/riscv/tlb.cc [hunk 9]
```diff
@@ -1054,37 +1054,47 @@ TLB::checkPermissions(STATUS status, PrivilegeMode pmode, Addr vaddr,
...
+            code = ExceptionCode::INSTG_PAGE;
+
+    } else {
+        if (mode == BaseMMU::Read)
+            code = ExceptionCode::LOAD_PAGE;
+        else if (mode == BaseMMU::Write)
+            code = ExceptionCode::STORE_PAGE;
+        else
+            code = ExceptionCode::INST_PAGE;
+    }
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc:1157]: Are if block and else block the same?

File: src/arch/riscv/tlb.cc [hunk 10]
```diff
@@ -1255,6 +1264,44 @@ TLB::L2TLBSendRequest(Fault fault, TlbEntry *e_l2tlb, const RequestPtr &req, Thr
...
+        }
+
+        if (mode == BaseMMU::hLdST) {
+            virt = 1;
+            two_stage_pmode = (PrivilegeMode)(RegVal)hstatus.spvp;
+        }
+    }
+    if (virt != 0) {
+        if (vsatp.mode == 0)
+            assert(0);
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc]: `assert (vsatp.mode != 0);`

File: src/arch/riscv/tlb.cc [hunk 11]
```diff
@@ -1255,6 +1264,44 @@ TLB::L2TLBSendRequest(Fault fault, TlbEntry *e_l2tlb, const RequestPtr &req, Thr
...
+    } else {
+        assert(0);
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc]: `asssert (virt != 0);`

File: src/cpu/o3/lsq_unit.cc [hunk 12]
```diff
@@ -1819,10 +1819,16 @@ LSQUnit::read(LSQRequest *request, ssize_t load_idx)
 
     if (!load_inst->isVector() && request->mainReq()->getSize() > 1 &&
         request->mainReq()->getVaddr() % request->mainReq()->getSize() != 0) {
-        DPRINTF(LSQUnit, "request: size: %u, Addr: %#lx, code: %d\n", request->mainReq()->getSize(),
-            request->mainReq()->getVaddr(), RiscvISA::ExceptionCode::LOAD_ADDR_MISALIGNED);
+        //DPRINTF(LSQUnit, "request: size: %u, Addr: %#lx, code: %d\n", request->mainReq()->getSize(),
+        //    request->mainReq()->getVaddr(), RiscvISA::ExceptionCode::LOAD_ADDR_MISALIGNED);
+        //return std::make_shared<RiscvISA::AddressFault>(request->mainReq()->getVaddr(),
+        //    RiscvISA::ExceptionCode::LOAD_ADDR_MISALIGNED);
```
Comment (shinezyy) [inline src/cpu/o3/lsq_unit.cc]: remove them?

File: src/arch/riscv/pagetable_walker.cc [hunk 13]
```diff
@@ -588,7 +590,7 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
         } else if (!pte.u) {
             doEndWalk = true;
             assert(0);
-        } else if (((mode == BaseMMU::Execute) || (mode == BaseMMU::hLdST)) && (!pte.x)) {
+        } else if (((mode == BaseMMU::Execute) || (isHInst)) && (!pte.x)) {
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc]: redundant () ?

File: src/arch/riscv/faults.cc [hunk 14]
```diff
@@ -165,10 +169,13 @@ RiscvFault::invoke(ThreadContext *tc, const StaticInstPtr &inst)
                   hstatus.spv = v;
                   if (v)
                       hstatus.spvp = pp;
-                  if (_code == INSTG_PAGE || _code == LOADG_PAGE || _code == STOREG_PAGE ||
+                  if (_code == INST_PAGE || _code == LOAD_PAGE || _code == STORE_PAGE ||
                       _code == LOAD_ADDR_MISALIGNED || _code == STORE_ADDR_MISALIGNED || _code == INST_ACCESS ||
                       _code == LOAD_ACCESS || _code == STORE_ACCESS)
                       tc->setMiscReg(MISCREG_HTVAL, 0);
+                  else if ((_code != INSTG_PAGE) && (_code != LOADG_PAGE) && (_code != STORE_PAGE))
+                      tc->setMiscReg(MISCREG_HTVAL, 0);
```
Comment (shinezyy) [inline src/arch/riscv/faults.cc:178]: This code block is large, consider to add { } to avoid confusion

File: src/arch/riscv/pagetable_walker.cc [hunk 15]
```diff
@@ -639,7 +632,11 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
             read = new Packet(request, MemCmd::ReadReq);
             read->allocate();
             DPRINTF(PageTableWalker, "Loading level%d PTE from %#x vaddr %#x\n", level, nextRead, entry.vaddr);
+        } else {
+            assert(0);
         }
+    } else {
+        assert(0);
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc]: consider to change above two assertions into assert (xxx conditions)

File: src/arch/riscv/tlb.cc [hunk 16]
```diff
@@ -1603,14 +1633,19 @@ TLB::translate(const RequestPtr &req, ThreadContext *tc,
...
-            if (two_stage_translation) {
-                if (vsatp.mode != 8 && hgatp.mode != 8)
-                    assert(0);
-                fault = doTwoStageTranslate(req, tc, translation, mode, delayed);
+            fault = misalignDataAddrCheck(req, mode);
+            if (fault == NoFault) {
+                two_stage_translation = hasTwoStageTranslation(tc, req, mode);
+                if (two_stage_translation) {
+                    if (vsatp.mode != 8 && hgatp.mode != 8)
+                        assert(0);
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc]: change to assert (xx condition)

File: src/arch/riscv/isa.cc [hunk 17]
```diff
@@ -410,6 +410,15 @@ ISA::readMiscReg(int misc_reg)
...
     }
+    if ((v == 1) && (misc_reg == MISCREG_SEPC)){
+        return readMiscRegNoEffect(MISCREG_VSEPC);
+    }
+    if ((v == 1) && (misc_reg == MISCREG_STVAL)){
+        return readMiscRegNoEffect(MISCREG_VSTVAL);
+    }
+    if ((v == 1) && (misc_reg == MISCREG_SCAUSE)){
+        return readMiscRegNoEffect(MISCREG_VSCAUSE);
+    }
```
Comment (shinezyy) [inline src/arch/riscv/isa.cc:414]: Can we rewrite above code like:

if (v == 1) and miscreg in [MISCREG_SATP, MISCREG_SPEC ....]:
   return ....

File: src/arch/riscv/pagetable_walker.cc [hunk 18]
```diff
@@ -615,7 +620,12 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
             vaddr_choose_flag = (PgBase & 0x3f) / 8;
             nextRead = (PgBase >> 6) << 6;
             gPaddr = nextRead;
+            if ((gPaddr & ~(((int64_t)1 << 41) - 1)) != 0) {
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc:721]: change assertion into assert(condition)

File: src/arch/riscv/pagetable_walker.cc [hunk 19]
```diff
@@ -697,12 +709,21 @@ Walker::WalkerState::twoStageWalk(PacketPtr &write)
                             gPaddr = ((pte.ppn << 12) & ~pg_mask) | (entry.vaddr & pg_mask & ~PGMASK);
                         }
                         gPaddr = gPaddr | (entry.vaddr & PGMASK);
+                        if ((gPaddr & ~(((int64_t)1 << 41) - 1)) != 0) {
+                            // this is a excep
+                            assert(0);
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc]: change assertion into assert(condition)

File: src/arch/riscv/pagetable_walker.cc [hunk 20]
```diff
@@ -712,18 +733,28 @@ Walker::WalkerState::twoStageWalk(PacketPtr &write)
...
                     fault = pageFault(true, false);
                     endWalk();
                 } else {
                     Addr shift = (PageShift + LEVEL_BITS * level);
                     Addr idx_f = (entry.vaddr >> shift) & LEVEL_MASK;
                     Addr idx = (idx_f >> L2TLB_BLK_OFFSET) << L2TLB_BLK_OFFSET;
                     gPaddr = (pte.ppn << PageShift) + (idx_f * l2tlbLineSize);
+                    if ((gPaddr & ~(((int64_t)1 << 41) - 1)) != 0) {
+                        // this is a excep
+                        assert(0);
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc]: change assertion into assert(condition)

File: src/arch/riscv/pagetable_walker.cc [hunk 21]
```diff
@@ -1552,19 +1594,33 @@ Walker::WalkerState::retry()
...
 {
+    Addr gpaddr = mainReq->getgPaddr();
+    Addr page_start = (entry.vaddr >> PageShift) << PageShift;
+    if (G && (gpaddrMode == 0) && isVsatp0Mode) {
+        Addr vaddr = 0;
+        if (r.req->isInstFetch()) {
+            if (r.req->getPC() < page_start) {
+                vaddr = page_start;
+            } else
+                vaddr = r.req->getPC();
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc:2020]: Either  use { } or not use { }. Don't mix them up in a code block.

XS-GEM5 always expected { } even if a block only has one line

File: src/arch/riscv/pagetable_walker.cc [hunk 22]
```diff
@@ -1129,6 +1153,28 @@ Walker::WalkerState::endWalk()
     delete read;
     read = NULL;
 }
+Fault
+Walker::WalkerState::startTwoStageWalkFromTLB(Addr ppn, Addr vaddr)
+{
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc:1455]: assert hgatp.mode == 8 ?

File: src/arch/riscv/pagetable_walker.cc [hunk 23]
```diff
@@ -637,10 +642,16 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
...
-            entry.paddr = (gPaddr >>12);
+            entry.paddr = gPaddr >> 12;
             entry.pte = pte;
-            entry.logBytes = PageShift + (level * LEVEL_BITS);
-            entry.level = level;
+            int put_level = 0;
+            if (twoStageLevel < level)
+                put_level = twoStageLevel;
+            else
+                put_level = level;
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc]: std::max?

File: src/cpu/o3/commit.cc [hunk 24]
```diff
@@ -1484,6 +1490,13 @@ Commit::commitHead(const DynInstPtr &head_inst, unsigned inst_num)
                     tid, head_inst->seqNum);
             return false;
         }
+        if (inst_fault->exception() == RiscvISA::ExceptionCode::LOAD_PAGE ||
+            inst_fault->exception() == RiscvISA::ExceptionCode::STORE_PAGE ||
+            inst_fault->exception() == RiscvISA::ExceptionCode::INST_PAGE ||
+            inst_fault->exception() == RiscvISA::ExceptionCode::INSTG_PAGE ||
+            inst_fault->exception() == RiscvISA::ExceptionCode::LOADG_PAGE ||
+            inst_fault->exception() == RiscvISA::ExceptionCode::STOREG_PAGE)
```
Comment (shinezyy) [inline src/cpu/o3/commit.cc]: put them in a set, then check whether exception is in this set?

File: src/cpu/o3/commit.cc [hunk 25]
```diff
@@ -1559,7 +1572,10 @@ Commit::commitHead(const DynInstPtr &head_inst, unsigned inst_num)
             auto exception_no =inst_fault->exception();
             if (exception_no == RiscvISA::ExceptionCode::LOAD_PAGE ||
                 exception_no == RiscvISA::ExceptionCode::STORE_PAGE ||
-                exception_no == RiscvISA::ExceptionCode::INST_PAGE) {
+                exception_no == RiscvISA::ExceptionCode::INST_PAGE ||
+                exception_no == RiscvISA::ExceptionCode::INSTG_PAGE ||
+                exception_no == RiscvISA::ExceptionCode::LOADG_PAGE ||
+                exception_no == RiscvISA::ExceptionCode::STOREG_PAGE) {
```
Comment (shinezyy) [inline src/cpu/o3/commit.cc]: simliar to above

File: src/arch/riscv/tlb.cc [hunk 26]
```diff
@@ -1082,6 +1079,26 @@ TLB::checkPermissions(STATUS status, PrivilegeMode pmode, Addr vaddr,
     return fault;
 }
 
+std::pair<bool, Fault>
+TLB::checkGPermissions(STATUS status,Addr vaddr,Addr gpaddr,BaseMMU::Mode mode, PTESv39 pte,bool h_inst){
+    bool continuePtw = false;
+    if (pte.v && !pte.r && !pte.w && !pte.x) {
+        assert(0);
+        continuePtw = true;
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc]: ???

File: src/arch/riscv/tlb.cc [hunk 27]
```diff
@@ -1274,6 +1291,111 @@ TLB::L2TLBSendRequest(Fault fault, TlbEntry *e_l2tlb, const RequestPtr &req, Thr
...
+    int hit_type = H_L1miss;
+
+    e[0] = lookup(vaddr, vsatp.asid, mode, false, true, allstage);
+    if (e[0]) {
+        hit_type = h_l1AllstageHit;
+        if (hgatp.vmid != e[0]->vmid)
+            assert(0);
+        if (vsatp.mode != 0) {
+            if ((mode == BaseMMU::Write && !e[0]->pteVS.d) || (!e[0]->pteVS.a))
+                fault = createPagefault(vaddr, 0, mode, false);
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc:1362]: add  { }

File: src/arch/riscv/tlb.cc [hunk 28]
```diff
@@ -1274,6 +1291,111 @@ TLB::L2TLBSendRequest(Fault fault, TlbEntry *e_l2tlb, const RequestPtr &req, Thr
...
+                gPaddr = pgBase |(gPaddr & PGMASK);
+
+                std::pair(continuePtw,fault) = checkGPermissions(status,vaddr,gPaddr,mode,e[0]->pte,req->get_h_inst());
+                if (fault != NoFault)
+                    return std::make_pair(hit_type,fault);
+
+                req->setPaddr(gPaddr);
+                return std::make_pair(hit_type,NoFault);
+            }
+            else{
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc]: } else {

File: src/arch/riscv/tlb.cc [hunk 29]
```diff
@@ -1274,6 +1291,111 @@ TLB::L2TLBSendRequest(Fault fault, TlbEntry *e_l2tlb, const RequestPtr &req, Thr
...
+                req->setTwoPtwWalk(true, 0, true);
+                req->setgPaddr(gPaddr);
+                return std::make_pair(hit_type,NoFault);
+            }
+        }
+        else{
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc]: } else {

File: src/arch/riscv/pagetable_walker.cc [hunk 30]
```diff
@@ -626,11 +679,45 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
...
+                        DPRINTF(PageTableWalker3, "level %d l2_level %d\n", level, l2_level);
+                        inl2Entry.paddr = l2pte.ppn;
+                        inl2Entry.pte = l2pte;
+                        if (l2_level == 0) {
+                            inl2Entry.index = (gPaddr >> (L2TLB_BLK_OFFSET + PageShift)) & L2TLB_L3_MASK;
+                            walker->tlb->L2TLBInsert(inl2Entry.gpaddr, inl2Entry, l2_level, L_L2L3, l2_i, false,
+                                                     gstage);
+                        }
+
+                        else if (l2_level == 1)  // hit level =1
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc]: add { }

File: src/arch/riscv/pagetable_walker.cc [hunk 31]
```diff
@@ -626,11 +679,45 @@ Walker::WalkerState::twoStageStepWalk(PacketPtr &write)
...
+                            walker->tlb->L2TLBInsert(inl2Entry.gpaddr, inl2Entry, l2_level, L_L2sp2, l2_i, false,
+                                                     gstage);
+                        else if (l2_level == 2)  //
+                            walker->tlb->L2TLBInsert(inl2Entry.gpaddr, inl2Entry, l2_level, L_L2sp1, l2_i, false,
+                                                     gstage);
```
Comment (shinezyy) [inline src/arch/riscv/pagetable_walker.cc:715]: add { }

File: src/arch/riscv/tlb.cc [hunk 32]
```diff
@@ -646,7 +654,12 @@ TLB::L2TLBInsertIn(Addr vpn, const TlbEntry &entry, int choose, EntryList *List,
             newEntry->isSquashed = true;
             stats.squashedInsertL2++;
             for (int i = 1; i < l2tlbLineSize; i++) {
-                newEntry = lookupL2TLB(vpn + step * i, entry.asid, BaseMMU::Read, true, choose, false);
+                if (translateMode == gstage)
+                    newEntry =
+                        lookupL2TLB(vpn + step * i, entry.vmid, BaseMMU::Read, true, choose, false, translateMode);
+                else
+                    newEntry =
+                        lookupL2TLB(vpn + step * i, entry.asid, BaseMMU::Read, true, choose, false, translateMode);
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc:664]: add { }

File: src/arch/riscv/tlb.cc [hunk 33]
```diff
@@ -838,53 +851,130 @@ TLB::demapPageL2(Addr vpn, uint64_t asid)
...
+                    l2TLBRemove(i, L_L2L2);
+            }
+            for (i = 0; i < l2TlbL3Size * l2tlbLineSize; i = i + l2tlbLineSize) {
+                if (tlbL2L3[i].trieHandle)
+                    l2TLBRemove(i, L_L2L3);
+            }
+            for (i = 0; i < l2TlbSpSize * l2tlbLineSize; i = i + l2tlbLineSize) {
+                if (tlbL2Sp[i].trieHandle)
+                    l2TLBRemove(i, L_L2sp1);
+            }
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc:888]: There should be a function, and call this function 4 times

File: src/arch/riscv/tlb.cc [hunk 34]
```diff
@@ -838,53 +851,130 @@ TLB::demapPageL2(Addr vpn, uint64_t asid)
...
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc]: There should be a function, and call this function 4 times

File: src/arch/riscv/tlb.cc [hunk 35]
```diff
@@ -838,53 +851,130 @@ TLB::demapPageL2(Addr vpn, uint64_t asid)
...
                 }
             }
+            if (tlbL2L3[i].trieHandle) {
+                if ((vpnl2l3 == 0 || (vpnl2l3 & l2l3_mask) == (tlbL2L3[i].gpaddr & l2l3_mask)) &&
+                    (asid == 0 || tlbL2L3[i].vmid == asid)) {
+                    l2TLBRemove(i, L_L2L3);
+                    DPRINTF(TLBVerbose3, "l2l3 remove vaddr %#x vpn %#x vpnl2l3\n", tlbL2L3[i].vaddr, vpn, vpnl2l3);
+                }
+            }
         }
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc:921]: There should be a function, and call this function x times.

File: src/arch/riscv/tlb.cc [hunk 36]
```diff
@@ -1415,6 +1725,13 @@ TLB::doTwoStageTranslate(const RequestPtr &req, ThreadContext *tc,
     bool continuePtw = false;
     int l1tlbtype = H_L1miss;
 
+    TLB *l2tlb;
+    if (isStage2)
+        l2tlb = this;
+    else
+        l2tlb = static_cast<TLB *>(nextLevel());
```
Comment (shinezyy) [inline src/arch/riscv/tlb.cc:1660]: add  { }

**Review by shinezyy — CHANGES_REQUESTED** [review]

File: src/arch/riscv/faults.cc (hunk 6 above)
Comment (shinezyy) [reply src/arch/riscv/faults.cc:117]: redundancy has not resolved yet @jueshiwenli ?

File: src/arch/riscv/tlb.cc (hunk 9 above)
Comment (shinezyy) [reply src/arch/riscv/tlb.cc:1157]: not resolved yet @jueshiwenli

**Review by shinezyy — APPROVED** [review]

**Pull request merged**